### ISSUE TITLE
pyomo.common.tempfiles.TempfileManager raises FileNotFoundError

### ISSUE THREAD

**Issue by jrlooker** [Pyomo/pyomo#1803]
I've recently started using pyomo.common.tempfiles.TempfileManager to set the temporary directory in a package, as in https://pyomo.readthedocs.io/en/stable/working_models.html#changing-the-temporary-directory. I was previously using TempfileManager from PyUtilib for about 4 years. 

My tests now fail. Here is some of the trace:

  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/base/solvers.py", line 571, in solve
    self._presolve(*args, **kwds)
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/solvers/plugins/solvers/CPLEX.py", line 349, in _presolve
    ILMLicensedSystemCallSolver._presolve(self, *args, **kwds)
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/solver/shellcmd.py", line 197, in _presolve
    OptSolver._presolve(self, *args, **kwds)
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/base/solvers.py", line 668, in _presolve
    self._convert_problem(args,
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/base/solvers.py", line 738, in _convert_problem
    return convert_problem(args,
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/base/convert.py", line 105, in convert_problem
    problem_files, symbol_map = converter.apply(*tmp, **tmpkw)
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/solvers/plugins/converter/model.py", line 72, in apply
    problem_filename = TempfileManager.\
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/common/tempfiles.py", line 67, in create_tempfile
    ans = tempfile.mkstemp(suffix=suffix, prefix=prefix, text=text, dir=dir)
  File "/home/user/anaconda38/lib/python3.8/tempfile.py", line 332, in mkstemp
    return _mkstemp_inner(dir, prefix, suffix, flags, output_type)
  File "/home/user/anaconda38/lib/python3.8/tempfile.py", line 250, in _mkstemp_inner
    fd = _os.open(file, flags, 0o600)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpc6s9g6hf/tmpjoxugx27.pyomo.lp'

If I run the tests individually on the commandline, then the FileNotFoundError does not occur.

If I run all the tests from a script, then the FileNotFoundError does occur.

If I run all the tests from the same script, but change the order of the tests, then the FileNotFoundError still occurs but during a different test.

Note that in all tests, I'm not acutally setting TempfileManager.tempdir. It appears in a method, but this method is not called during these tests. So just the import "from pyomo.common.tempfiles import TempfileManager" is being run.

Now if I change my code so that "TempfileManager.tempdir = None" is always called for each test, then the FileNotFoundError no longer occurs.

Can you help?

I'm using Python 3.8.5 from Anaconda, Pyomo 5.7.2 and Ubuntu 18.04 LTS.

Thanks,
Jason

**Comment by mrmundt** [comment]
Thanks for reporting this. Is there any chance you could either paste your script or point me to where it's located so I can try to reproduce the error?

**Comment by mrmundt** [comment]
@jrlooker - Actually, I'm fairly certain I found the issue and have opened a PR for it.

**Comment by jrlooker** [comment]
Sorry it has taken me a while to get back to you. Thanks for fixing this so quickly!

**Comment by jrlooker** [comment]
Unfortunately, after upgrading to Pyomo 5.7.3 (using pip install --upgrade pyomo) and running my tests, I still obtain the same behaviour described above.

Sorry, I can't provide you with any of my code. 

After more investigation, I may know what the problem is:

Recall that the issue occurs when I run a test script, which itself imports an number of other tests and runs "test1.main(), test2.main(), ... " . If I run each of these sub-tests in their own scripts, then the issue does not occur. I believe the issue is, in some of these sub-tests I'm using multiprocessing.Pool to run a number of Pyomo models in parallel. The first sub-test that is run in parallel passes, but the next sub-test fails.

Note that a method is called both in the main process and all the workers (using the initializer kwarg of Pool) to set TempfileManager.tempdir:

def set_tempdir(self):
    if self.tempdir:
        TempfileManager.tempdir = self.tempdir

However, self.tempdir = None in these tests.

Again, recall that if I force TempfileManager.tempdir = None, then the issue does not occur.

Here is some of the output from Pyomo 5.7.3:

  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/base/solvers.py", line 575, in solve
    self._presolve(*args, **kwds)
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/solvers/plugins/solvers/CPLEX.py", line 349, in _presolve
    ILMLicensedSystemCallSolver._presolve(self, *args, **kwds)
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/solver/shellcmd.py", line 198, in _presolve
    OptSolver._presolve(self, *args, **kwds)
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/base/solvers.py", line 672, in _presolve
    self._convert_problem(args,
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/base/solvers.py", line 742, in _convert_problem
    return convert_problem(args,
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/opt/base/convert.py", line 105, in convert_problem
    problem_files, symbol_map = converter.apply(*tmp, **tmpkw)
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/solvers/plugins/converter/model.py", line 72, in apply
    problem_filename = TempfileManager.\
  File "/home/user/anaconda38/lib/python3.8/site-packages/pyomo/common/tempfiles.py", line 66, in create_tempfile
    ans = tempfile.mkstemp(suffix=suffix, prefix=prefix, text=text, dir=dir)
  File "/home/user/anaconda38/lib/python3.8/tempfile.py", line 332, in mkstemp
    return _mkstemp_inner(dir, prefix, suffix, flags, output_type)
  File "/home/user/anaconda38/lib/python3.8/tempfile.py", line 250, in _mkstemp_inner
    fd = _os.open(file, flags, 0o600)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmp0fepkix2/tmput6kcger.pyomo.lp'

Can you help?

**Comment by mrmundt** [comment]
Darn. The odd thing about this situation is that upon initialization, `TempfileManager.tempdir = None`:
```
In [1]: from pyomo.common.tempfiles import TempfileManager
In [2]: t = TempfileManager
In [3]: print(t.tempdir)
None
```
so your step to force it to be `None` is repetitive. I can't recreate the issue on my end, so I'm unsure as to why your test in particular is hitting this.

**Comment by jrlooker** [comment]
A couple more clues:
1. The part of my code that uses TempfileManager is about 4 years old and this problem only began occurring when TempfileManager moved to pyomo.common.tempfiles, so could you see if any change occurred that could impact parallel computation?
2. In my tests, I'm using tempfile.TemporaryDirectory to store the output from my code. Could this be interacting with pyomo.common.tempfiles?

Running parallel processes in Python can be tricky. I've been careful to enclose all my code under one function, which is then mapped over input data using multiprocessing.Pool.map. Unfortunately, TempfileManager is like a global variable and needs to be handled separately using the Pool initializer. If TempfileManager now has some "non-local" dependencies, then this may break parallel computation using process Pools.

I'll try and reproduce the problem in some toy code, but I'm not optimistic that I can.

**Comment by jrlooker** [comment]
I've managed to write some toy code that reproduces this issue.

Firstly, the issue has nothing to do with parallel processing.

To summarise, TempfileManager.tempdir remembers its value between calls to
functions, which depend on TempfileManager.tempdir, and are defined in
different modules that are themselves imported to the same module. The issue
manifested itself for me because I'm setting TempfileManager.tempdir to a
temporary directory created, and deleted, by tempfile.TemporaryDirectory.

The attached scripts demonstrate the issue:
    * Run ./test1.py on the command line and it will run without any issues.
    * Run ./test2.py on the command line and it will run without any issues.
    * Run ./runall.py on the command line and it will reproduce the issue.

[runall.txt](https://github.com/Pyomo/pyomo/files/6033431/runall.txt)
[test1.txt](https://github.com/Pyomo/pyomo/files/6033432/test1.txt)
[test2.txt](https://github.com/Pyomo/pyomo/files/6033433/test2.txt)

These scripts demonstrate why setting TempfileManager.tempdir = None, even when
it isn't needed, fixes the issue, as it replaces its previous assigned value.

**Comment by mrmundt** [comment]
Thanks, @jrlooker - I'll work on this.

**Comment by jsiirola** [comment]
Looking at the submitted example, I believe that things are working as designed / intended.  In the example above, `test1` creates a temporary directory using a context manager.  Within the context manager's scope, the current TempfileManager's `tempdir` is set to that directory.  When the test exits the context manager, `tempfile.TemporaryDirectory` correctly deletes the directory.  However, there is nothing to tell the `TempfileManager` that the directory has fallen out of scope, so the next attempt to create a temporary file in that directory (correctly) fails.

The crux of the problem is that you are using two systems for managing temporary objects that do not (and really can not) communicate with each other.  A better solution might be to either hand-roll everything (using the `tempfile` module), or else leverage `pyomo.common.tempfiles.TempfileManager`'s context manager capabilities.  For example, I think the following should work for your use case:

```python
import glob, os, time
from multiprocessing import Pool
from pyomo.common.tempfiles import TempfileManager

def run_test(test_name):
    if test_name == 'final':
        time.sleep(1) # sleep so other jobs are "for sure" done
        print("tempdir:", tempdir)
        print("files:", list(glob.glob(os.path.join(tempdir, '*'))))
    else:
        TempfileManager.create_tempfile(prefix=test_name + "_", suffix='.txt')

with TempfileManager as context:
    # Note that the temp directory that we will (by default) use for
    # creating temporary files is managed by the context: so it will not
    # be deleted until the context itself goes out of scope.
    context.tempdir = tempdir = context.create_tempdir()
    with Pool(5) as p:
        p.map(run_test, ['test1', 'test2', 'test3', 'final'])

print("tempdir exists:", os.path.exists(tempdir))
```
when run:
```
tempdir: /tmp/tmp47np9qip
files: ['/tmp/tmp47np9qip/test1_6aq0aldn.txt', '/tmp/tmp47np9qip/test2_3e8whift.txt', '/tmp/tmp47np9qip/test3_9_hyr6gn.txt']
tempdir exists: False
```
The trick here is that the context that creates / manages the temporary directory is *outside* the multiprocessing loop.

**Comment by jrlooker** [comment]
Thanks, John

I’ll give that a go. 

Cheers,
Jason
